Refactor paths

ustulation opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/maidsafe/config_file_handler/29)

<!-- Reviewable:end -->

ustulation committed:
refactor/files: ask for writability
Ask users for specifying writablitiy check rather than internally assume it, so that it covers more use cases and gives more flexibility.

ustulation committed:
fix/paths: fix various search paths
Fix search paths for file for various OS.

maidsafe-highfive commented:
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @Fraser999 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

ustulation committed:
fix/mutex: slap global mutex
Apply global mutexing for file creation and deletion as multiple test threads take a stab at it.

ustulation committed:
chore/file: add missing file

ustulation committed:
feat/create-dir: create non-existent dirs

afck commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/file_handler.rs, line 275 \[r2\]](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29#-KOQ4ZuOgzmSJAgxU8IC:-KOQ4ZuPKNTg5ftVEL4v:b-bq5quu) ([raw file](https://github.com/maidsafe/config_file_handler/blob/6a439b25a8dad761fac818d8e2aee6b7df3c6147/src/file_handler.rs#L275)):*
> ```Rust
>                 .ok_or(Error::Io(io::Error::new(io::ErrorKind::NotFound, "Current bin dir"))))
>             .to_path_buf();
>     }
> ```

Should it really always go up two levels on OS X? E.g. for testing, it would be surprising if the config files were in `Resources` instead of `target/release`, I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29)*
<!-- Sent from Reviewable.io -->

ustulation committed:
test/directory: print binary resource dir

ustulation committed:
fix/function: fix windows error

ustulation commented:
*[src/file_handler.rs, line 275 \[r2\]](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29#-KOQ4ZuOgzmSJAgxU8IC:-KOQIiX4AdDkYceuKt5M:b-pxh5nb) ([raw file](https://github.com/maidsafe/config_file_handler/blob/6a439b25a8dad761fac818d8e2aee6b7df3c6147/src/file_handler.rs#L275)):*
<details><summary><i>Previously, afck (Andreas Fackler) wrote…</i></summary>
> Should it really always go up two levels on OS X? E.g. for testing, it would be surprising if the config files were in `Resources` instead of `target/release`, I think.
</details>
It is split into 2 current_bin_resource_dir and current_bin_dir - maybe some sync loss between a commit and your comment ? Or did you mean something else ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29)*
<!-- Sent from Reviewable.io -->

ustulation commented:
*[src/file_handler.rs, line 275 \[r2\]](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29#-KOQ4ZuOgzmSJAgxU8IC:-KOQJaNICLAawaCmQaHA:bsxoaov) ([raw file](https://github.com/maidsafe/config_file_handler/blob/6a439b25a8dad761fac818d8e2aee6b7df3c6147/src/file_handler.rs#L275)):*
<details><summary><i>Previously, ustulation (Spandan Sharma) wrote…</i></summary>
> It is split into 2 current_bin_resource_dir and current_bin_dir - maybe some sync loss between a commit and your comment ? Or did you mean something else ?
</details>
(Clarifying more): this https://github.com/maidsafe/config_file_handler/pull/29/commits/f86807e17404f8db472b99912a1370fa6d66e6ab before your comment split that into two functions

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29)*
<!-- Sent from Reviewable.io -->

afck commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/file_handler.rs, line 275 \[r2\]](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29#-KOQ4ZuOgzmSJAgxU8IC:-KOQK2mKVC5L_GMsPVja:b-3w5yuf) ([raw file](https://github.com/maidsafe/config_file_handler/blob/6a439b25a8dad761fac818d8e2aee6b7df3c6147/src/file_handler.rs#L275)):*
<details><summary><i>Previously, ustulation (Spandan Sharma) wrote…</i></summary>
> (Clarifying more): this https://github.com/maidsafe/config_file_handler/pull/29/commits/f86807e17404f8db472b99912a1370fa6d66e6ab before your comment split that into two functions
</details>
Ah, sorry, I hadn't realised that `current_bin_dir` is tried first. :+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/config_file_handler/29)*
<!-- Sent from Reviewable.io -->

ustulation committed:
fix/windows: fix more compile error

ustulation committed:
refactor/bundle-dir: check if valid bundle dir

ustulation committed:
refactor/mac: avoid bundle resource path
Avoid bundle resource path while creating a file. This path must be preferred to be read from but should never be attempted to written to.
